### PR TITLE
[Road to V6] Permutations & Packing

### PR DESCRIPTION
**Pull request by qantik**
This pull request adds two new features to Lattigo:

 - (27cb1b2): More efficient ring packing algorithm inspired by [HERMES](https://eprint.iacr.org/2023/1244):
 - (a4e50f6): Slot-wise permutations based on the existing linear transformation feature.

### New Ring Packing API:

- New Evaluator: `he.RingPackingEvaluator`:
    - `NewRingPackingEvaluator(evk *RingPackingEvaluationKey)`
    - `Extract(ct *rlwe.Ciphertext, idx []int, naive bool) (cts map[int]*rlwe.Ciphertext, err error)`
    - `Repack(cts map[int]*rlwe.Ciphertext, naive bool) (ct *rlwe.Ciphertext, err error)`
    - `Split(ctN, ctEvenNHalf, ctOddNHalf *rlwe.Ciphertext) (err error)`
    - `Merge(ctEvenNHalf, ctOddNHalf, ctN *rlwe.Ciphertext) (err error)`
    - `ShallowCopy() *RingPackingEvaluator`
- New EvaluationKey: `he.RingPackingEvaluationKey`

### CODE REVIEW

**Comment by Pro7ech** [comment]
@qantik my commit history was erased with your latest forcepush

**Comment by Pro7ech** [comment]
@qantik The latest force push erased the commit history and my co-authorship

File: he/ring_packing_test.go [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+	require.NoError(t, err)
+
+	// Ensures that ct is encrypted at the max
+	// defined ring degree
+	require.Equal(t, ct.LogN(), eval.MaxLogN())
+
+	// Generates a random index selection
+	// of size N/2 (to test that omitted
+	// elements output zero coefficients)
+	r := rand.New(rand.NewSource(0))
```
Comment (romainbou) [inline he/ring_packing_test.go:449]: Same comment for the use of `math/rand` and GoSec

**Comment by qantik** [comment]
The ignored `gosec` exception at `utils/structs/vector.go:276` appears to be a bug in `gosec` itself.
No matter how and where the exception comment string is placed it is ignored by the tool.

**Review by romainbou — APPROVED** [review]

**Pull request merged**